crates/compose/src/lib.rs

```diff
@@ -98,19 +148,19 @@ struct Composer<'a, L> {
 }
 
 impl<'a, L: ComponentSourceLoader> Composer<'a, L> {
-    async fn compose(mut self, component: &LockedComponent) -> Result<Vec<u8>, ComposeError> {
+    async fn compose(mut self, component: &L::Component) -> Result<Vec<u8>, ComposeError> {
         let source = self
             .loader
-            .load_component_source(&component.source)
+            .load_component_source(component)
             .await
             .map_err(ComposeError::PrepareError)?;
 
-        if component.dependencies.is_empty() {
+        if component.dependencies().len() == 0 {
```

Comment:
`is_empty()` is nightly-only on ExactSizeIterator (at least in the version of Rust I'm using).  To answer your next question, I changed to ExactSizeIterator because different implementations of `.dependencies()` produce different types (manifest vs lockfile).  But I can collect to Vec or whatever if you prefer.